server/src/main/java/org/elasticsearch/action/support/replication/PendingReplicationActions.java

```diff
@@ -0,0 +1,98 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.support.replication;
+
+import org.elasticsearch.action.UnavailableShardsException;
+import org.elasticsearch.action.support.RetryableAction;
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.index.shard.IndexShardClosedException;
+import org.elasticsearch.index.shard.ReplicationGroup;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+
+public class PendingReplicationActions implements Consumer<ReplicationGroup>, Releasable {
+
+    private final Map<String, Set<RetryableAction<?>>> onGoingReplicationActions = ConcurrentCollections.newConcurrentMap();
+    private final ShardId shardId;
+    private final ThreadPool threadPool;
+
+    public PendingReplicationActions(ShardId shardId, ThreadPool threadPool) {
+        this.shardId = shardId;
+        this.threadPool = threadPool;
+    }
+
+    public void addPendingAction(String allocationId, RetryableAction<?> replicationAction) {
+        Set<RetryableAction<?>> ongoingActionsOnNode = onGoingReplicationActions.get(allocationId);
+        if (ongoingActionsOnNode != null) {
+            ongoingActionsOnNode.add(replicationAction);
+            if (onGoingReplicationActions.containsKey(allocationId) == false) {
+                replicationAction.cancel(new UnavailableShardsException(shardId,
+                    "Replica unavailable - replica could have left ReplicationGroup or IndexShard might have closed"));
+            }
+        } else {
+            replicationAction.cancel(new UnavailableShardsException(shardId,
+                "Replica unavailable - replica could have left ReplicationGroup or IndexShard might have closed"));
+        }
+    }
+
+    public void removeReplicationAction(String allocationId, RetryableAction<?> action) {
+        Set<RetryableAction<?>> ongoingActionsOnNode = onGoingReplicationActions.get(allocationId);
+        if (ongoingActionsOnNode != null) {
+            ongoingActionsOnNode.remove(action);
+        }
+    }
+
+    @Override
+    public synchronized void accept(ReplicationGroup replicationGroup) {
+        Set<String> trackedAllocationIds = replicationGroup.getTrackedAllocationIds();
+
+        for (String targetAllocationId : trackedAllocationIds) {
+            onGoingReplicationActions.putIfAbsent(targetAllocationId, ConcurrentCollections.newConcurrentSet());
+        }
+        ArrayList<Set<RetryableAction<?>>> toCancel = new ArrayList<>();
+        for (String allocationId : onGoingReplicationActions.keySet()) {
```

Comment:
`onGoingReplicationActions.keySet()` is only eventually consistent (i.e. can race against `addPendingAction`).